components/filter/filter.js

```diff
@@ -0,0 +1,259 @@
+					<d2l-menu label="${this.localize('components.filter.filters')}">
+						<slot
+							@d2l-filter-dimension-data-change="${this._handleDataChangeListDimension}"
+							@d2l-filter-dimension-value-data-change="${this._handleDataChangeListDimensionValue}"
+							@d2l-filter-dimension-slot-change="${this._handleSlotChange}"
+							@slotchange="${this._handleSlotChange}"
+						></slot>
+						${dimensions}
+					</d2l-menu>
+				</d2l-dropdown-menu>
+			</d2l-dropdown-button-subtle>
+		`;
+	}
+
+	_buildDimension(dimension) {
+		let dimensionHTML;
+		switch (dimension.tagName.toLowerCase()) {
+			case 'd2l-filter-dimension':
+				dimensionHTML = this._createListDimension(dimension);
+				break;
+		}
+		return html`
+			<d2l-hierarchical-view
+				@d2l-hierarchical-view-show-complete="${this._handleDimensionShowComplete}"
+				@d2l-hierarchical-view-show-start="${this._handleDimensionShowStart}"
+				id="${dimension.key}">
+
+				${dimensionHTML}
+			</d2l-hierarchical-view>
+		`;
+	}
+
+	_buildDimensions() {
+		return Array.from(this._dimensions.values(), (dimension) => {
+			const builtDimension = this._buildDimension(dimension);
+			return html`<d2l-menu-item text="${dimension.text}">
+				${builtDimension}
+			</d2l-menu-item>`;
+		});
+	}
+
+	_createListDimension(dimension) {
+		const slot = dimension.shadowRoot.querySelector('slot');
+		const items = slot.assignedNodes().filter(node => node.nodeType ===  Node.ELEMENT_NODE);
+		return html`
+			<d2l-list
+				@d2l-list-selection-change="${this._handleChangeListDimension}"
+				extend-separators>
+
+				${items.map(item => html`
+					<d2l-list-item
+						key="${item.key}"
+						selectable
+						?selected="${item.selected}"
+						slim>
+
+						<div class="d2l-filter-dimension-value-text">${item.text}</div>
+					</d2l-list-item>
+				`)}
+			</d2l-list>
+		`;
+	}
+
+	_dispatchChangeEvent(eventDetail) {
+		this.dispatchEvent(new CustomEvent('d2l-filter-change', { bubbles: false, composed: false, detail: eventDetail }));
+	}
+
+	_handleChangeListDimension(e) {
+		const dimensionKey = e.composedPath()[0].parentNode.id;
+		const valueKey = e.detail.key;
+		const selected = e.detail.selected;
+
+		// Update the corresponding d2l-filter-dimension-value to keep them in sync
+		const dimension = this._dimensions.get(dimensionKey);
+		const slot = dimension.shadowRoot.querySelector('slot');
+		const items = slot.assignedNodes().filter(node => node.nodeType ===  Node.ELEMENT_NODE);
+		const item = items.find(item => item.key === valueKey);
+		item.selected = selected;
+
+		this._dispatchChangeEvent({ dimension: dimensionKey, value: { key: valueKey, selected: selected } });
+	}
+
+	_handleDataChangeListDimension(e) {
+		const dimension = this.shadowRoot.getElementById(e.target.key);
+		const changes = e.detail.changes;
+		if (changes.has('text')) {
+			const menuItem = dimension.parentNode;
+			if (menuItem.text !== changes.get('text')) {
+				this.requestUpdate();
```

Comment:
If a dimension's text changes, there's a number of places we'd need to update, so we just call a re-render here.  But we do try to determine if we need to first.  This logic is fairly specific, and will be specific for each data value and dimension type (you can see more examples in `_handleDataChangeListDimensionValue` directly below).  I was struggling to determine how far to go with this.  For this review I went all the way (determine if we need to change a value, and then try to change it without a re-render if there are no bad consequences).  We could also:
1) Try to determine if we need it, and then just always re-render
2) Try to determine if we need it if it's easy/reusable, and then re-render
3) Just always re-render

As we move down the list the logic will get much simpler, at the cost of re-rendering.  Thoughts?